# Data Transformation and Manipulation

## Objective 
Data is rarely in the perfect format for analysis. We often need to restructure, clean, and process it before we can derive meaningful insights. In this tutorial, we'll explore essential techniques for data transformation and manipulation in Python.

## Pre-requisites: 

- Python Environment: You should have a Python environment set up on your system. If you don't have Python installed, you can download it from the [official Python website](https://www.python.org/downloads/). We recommend downloading Python 3.8 or above.
- Jupyter Notebook: This code is intended to be run in a Jupyter Notebook environment. Make sure you have [Jupyter Notebook installed](https://jupyter.org/install).
- Library Installation: We'll be using the Pandas library, which is incredibly powerful for data manipulation. You can install it using the following command:
 
To install, run the following commands:
"**!pip3 install pandas**"

In [1]:
# Import the necessary libraries
import pandas as pd

# Load a the dataset
data = pd.read_csv('titatnic.csv')

Before we dive into transformation, it's essential to understand the data we're working with. Let's take a look at the first few rows and get a summary of the dataset.

In [4]:
# View the first few rows of the dataset
data.head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S

Here, we view the first few rows of the loaded dataset using the head() method to provide an initial look at the data.

In [5]:
# Get a summary of the dataset
summary = data.describe()
print(summary)

       PassengerId    Survived      Pclass         Age       SibSp  \
count   418.000000  418.000000  418.000000  332.000000  418.000000   
mean   1100.500000    0.363636    2.265550   30.272590    0.447368   
std     120.810458    0.481622    0.841838   14.181209    0.896760   
min     892.000000    0.000000    1.000000    0.170000    0.000000   
25%     996.250000    0.000000    1.000000   21.000000    0.000000   
50%    1100.500000    0.000000    3.000000   27.000000    0.000000   
75%    1204.750000    1.000000    3.000000   39.000000    1.000000   
max    1309.000000    1.000000    3.000000   76.000000    8.000000   

            Parch        Fare  
count  418.000000  417.000000  
mean     0.392344   35.627188  
std      0.981429   55.907576  
min      0.000000    0.000000  
25%      0.000000    7.895800  
50%      0.000000   14.454200  
75%      0.000000   31.500000  
max      9.000000  512.329200  


Now we generate a summary of the dataset using the describe() method and print statistics such as count, mean, standard deviation, minimum, and maximum values for numerical columns in the dataset. Let's break down the output:

- **count**: This shows the count of non-null values in each column. For example, there are 418 non-null values in 'PassengerId' and 'Pclass', but there are only 332 non-null values in the 'Age' column, indicating that there are missing values in the 'Age' column.
- **mean**: This displays the mean (average) value for each column. For instance, the mean 'PassengerId' is 1100.5, the mean 'Survived' is approximately 0.364, and the mean 'Age' is roughly 30.27.
- **std**: This represents the standard deviation for each column. Standard deviation is a measure of the spread or dispersion of the values. For example, the standard deviation of 'PassengerId' is approximately 120.81, and the standard deviation of 'Age' is roughly 14.18.
- **min**: This shows the minimum value in each column. It indicates the smallest value present in each column. For example, the minimum 'PassengerId' is 892, and the minimum 'Age' is 0.17.
- **25%**: This displays the 25th percentile of the data, which is also known as the first quartile. It indicates the value below which 25% of the data falls. For instance, the 25th percentile for 'PassengerId' is 996.25, and for 'Age', it's 21.
- **50%**: This represents the median, which is the middle value in the dataset when it's sorted in ascending order. It's also known as the 50th percentile. For example, the median 'PassengerId' is 1100.5, and the median 'Age' is 27.
- **75%**: This row displays the 75th percentile of the data, which is also known as the third quartile. It indicates the value below which 75% of the data falls. For example, the 75th percentile for 'PassengerId' is 1204.75, and for 'Age', it's 39.
- **max**: This row shows the maximum value in each column. It indicates the largest value present in each column. For instance, the maximum 'PassengerId' is 1309, and the maximum 'Age' is 76.

Now, we will filter and subset data to focus on specific rows or columns that are relevant to our analysis. We will select a single column ('Age'), filtering the dataset to include only passengers older than 30, and printing the resulting data frame.

In [6]:
# Select a single column
column_data = data['Age']

# Filter rows based on a condition (e.g., passengers who are older than 30)
filtered_data = data[data['Age'] > 30]

# Print a message when the condition is true
if not filtered_data.empty:
    print("Passengers older than 30 years:")
    print(filtered_data)
else:
    print("No passengers older than 30 years found.")

Passengers older than 30 years:
     PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
11           903         0       1   
13           905         0       2   
..           ...       ...     ...   
404         1296         0       1   
407         1299         0       1   
411         1303         1       1   
414         1306         1       1   
415         1307         0       3   

                                                Name     Sex   Age  SibSp  \
0                                   Kelly, Mr. James    male  34.5      0   
1                   Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
2                          Myles, Mr. Thomas Francis    male  62.0      0   
11                        Jones, Mr. Charles Cresson    male  46.0      0   
13                              Howard, Mr. Benjamin    male  63.0      1   
..                                               

Now, we combine multiple conditions to filter the dataset and display passengers older than 30 years in the feature Pclass 1.

In [9]:
# Combine multiple conditions (use '&' for 'and', '|' for 'or')
combined_filter = data[(data['Age'] > 30) & (data['Pclass'] == 1)]

# Print a message when the condition is true
if not combined_filter.empty:
    print("Passengers older than 30 years in Pclass 1:")
    print(combined_filter)
else:
    print("No passengers older than 30 years in Pclass 1 found.")

Passengers older than 30 years in Pclass 1:
     PassengerId  Survived  Pclass  \
11           903         0       1   
14           906         1       1   
20           912         0       1   
24           916         1       1   
28           920         0       1   
..           ...       ...     ...   
397         1289         1       1   
404         1296         0       1   
407         1299         0       1   
411         1303         1       1   
414         1306         1       1   

                                                  Name     Sex   Age  SibSp  \
11                          Jones, Mr. Charles Cresson    male  46.0      0   
14   Chaffee, Mrs. Herbert Fuller (Carrie Constance...  female  47.0      1   
20                              Rothschild, Mr. Martin    male  55.0      1   
24     Ryerson, Mrs. Arthur Larned (Emily Maria Borie)  female  48.0      1   
28                             Brady, Mr. John Bertram    male  41.0      0   
..                       

Dealing with missing data is a common challenge. So, we calculate and print the number of missing values in each column of the dataset using isnull().sum().

In [11]:
# Check for missing values
missing_values = data.isnull().sum()
print(missing_values)

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64


In [12]:
# Remove rows with missing values
data_cleaned = data.dropna()
print(data_cleaned)

     PassengerId  Survived  Pclass  \
12           904         1       1   
14           906         1       1   
24           916         1       1   
26           918         1       1   
28           920         0       1   
..           ...       ...     ...   
404         1296         0       1   
405         1297         0       2   
407         1299         0       1   
411         1303         1       1   
414         1306         1       1   

                                                  Name     Sex   Age  SibSp  \
12       Snyder, Mrs. John Pillsbury (Nelle Stevenson)  female  23.0      1   
14   Chaffee, Mrs. Herbert Fuller (Carrie Constance...  female  47.0      1   
24     Ryerson, Mrs. Arthur Larned (Emily Maria Borie)  female  48.0      1   
26                        Ostby, Miss. Helene Ragnhild  female  22.0      0   
28                             Brady, Mr. John Bertram    male  41.0      0   
..                                                 ...     ...   ... 

In [13]:
# Fill missing values with a specific value
data_filled = data.fillna(0)
print(data_filled)

     PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
3            895         0       3   
4            896         1       3   
..           ...       ...     ...   
413         1305         0       3   
414         1306         1       1   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                Kelly, Mr. James    male  34.5      0      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                       Myles, Mr. Thomas Francis    male  62.0      0      0   
3                                Wirz, Mr. Albert    male  27.0      0      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
..                                            ...     ...

Sometimes, we need to create new features to better represent our data. To do that, we:
- Define a function **categorize_age**  to categorize passengers into 'Child', 'Adult', or 'Senior' based on their age.
- Then we apply this function to the 'Age' column to create a new 'Age_Category' column in the dataset.
- The resulting data frame contains the 'Age_Category' column indicating the age category for each passenger.

In [9]:
# Create a new feature based on existing ones. Define a function to categorize age

def categorize_age(age):
    if age < 18:
        return 'Child'
    elif age < 65:
        return 'Adult'
    else:
        return 'Senior'

# Apply the function to create the 'Age_Category' column
data['Age_Category'] = data['Age'].apply(categorize_age)

# Display the updated data frame
data

PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
3            895         0       3   
4            896         1       3   
..           ...       ...     ...   
413         1305         0       3   
414         1306         1       1   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                Kelly, Mr. James    male  34.5      0      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                       Myles, Mr. Thomas Francis    male  62.0      0      0   
3                                Wirz, Mr. Albert    male  27.0      0      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
..                                            ...     ...   ...    ...    ...   
413                            Spector, Mr. Woolf    male   NaN      0      0   
414                  Oliva y Ocana, Dona. Fermina  female  39.0      0      0   
415                  Saether, Mr. Simon Sivertsen    male  38.5      0      0   
416                           Ware, Mr. Frederick    male   NaN      0      0   
417                      Peter, Master. Michael J    male   NaN      1      1   

                 Ticket      Fare Cabin Embarked Age_Category  
0                330911    7.8292   NaN        Q        Adult  
1                363272    7.0000   NaN        S        Adult  
2                240276    9.6875   NaN        Q        Adult  
3                315154    8.6625   NaN        S        Adult  
4               3101298   12.2875   NaN        S        Adult  
..                  ...       ...   ...      ...          ...  
413           A.5. 3236    8.0500   NaN        S       Senior  
414            PC 17758  108.9000  C105        C        Adult  
415  SOTON/O.Q. 3101262    7.2500   NaN        S        Adult  
416              359309    8.0500   NaN        S       Senior  
417                2668   22.3583   NaN        C       Senior  

[418 rows x 13 columns]

For another example of creating a feature,  We created new features, **Family_Size** and **Title,** which can be used for various analyses and can provide valuable insights. For example, "Family_Size" might help understand the dynamics of families traveling together on the Titanic, and "Title" could be used to analyze passenger demographics based on their social status or gender.

In [10]:
# Create a new feature based on existing ones (e.g., family size)
data['Family_Size'] = data['SibSp'] + data['Parch']

# Apply a function to create a new feature (e.g., extract the title from the 'Name' column)
data['Title'] = data['Name'].apply(lambda name: name.split(',')[1].split('.')[0].strip())
data

PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
3            895         0       3   
4            896         1       3   
..           ...       ...     ...   
413         1305         0       3   
414         1306         1       1   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                Kelly, Mr. James    male  34.5      0      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                       Myles, Mr. Thomas Francis    male  62.0      0      0   
3                                Wirz, Mr. Albert    male  27.0      0      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
..                                            ...     ...   ...    ...    ...   
413                            Spector, Mr. Woolf    male   NaN      0      0   
414                  Oliva y Ocana, Dona. Fermina  female  39.0      0      0   
415                  Saether, Mr. Simon Sivertsen    male  38.5      0      0   
416                           Ware, Mr. Frederick    male   NaN      0      0   
417                      Peter, Master. Michael J    male   NaN      1      1   

                 Ticket      Fare Cabin Embarked Age_Category  Family_Size  \
0                330911    7.8292   NaN        Q        Adult            0   
1                363272    7.0000   NaN        S        Adult            1   
2                240276    9.6875   NaN        Q        Adult            0   
3                315154    8.6625   NaN        S        Adult            0   
4               3101298   12.2875   NaN        S        Adult            2   
..                  ...       ...   ...      ...          ...          ...   
413           A.5. 3236    8.0500   NaN        S       Senior            0   
414            PC 17758  108.9000  C105        C        Adult            0   
415  SOTON/O.Q. 3101262    7.2500   NaN        S        Adult            0   
416              359309    8.0500   NaN        S       Senior            0   
417                2668   22.3583   NaN        C       Senior            2   

      Title  
0        Mr  
1       Mrs  
2        Mr  
3        Mr  
4       Mrs  
..      ...  
413      Mr  
414    Dona  
415      Mr  
416      Mr  
417  Master  

[418 rows x 15 columns]

Aggregation and grouping are essential for summarizing data. We can group data by a specific column and perform operations on those groups. Below, we will group the data by the 'Pclass' column and calculate the mean age for each passenger class.

In [26]:
# Group data by the 'Pclass' column and calculate the mean of 'Age'
grouped_data = data.groupby('Pclass')['Age'].mean()
print("Mean Age by Pclass:")
print(grouped_data)

Mean Age by Pclass:


Pclass
1    40.918367
2    28.777500
3    24.027945
Name: Age, dtype: float64

We first group the data by the 'Pclass' column and calculate the mean age for each passenger class. Then, we group the data by 'Pclass' and calculate multiple statistics (mean, median, and standard deviation) for the 'Fare' column. The results are displayed for each grouping.

In [27]:
# Calculate multiple statistics (mean, median, and standard deviation) for 'Fare' by 'Pclass'
agg_data = data.groupby('Pclass')['Fare'].agg(['mean', 'median', 'std'])
print("\nStatistics for Fare by Pclass:")
print(agg_data)


Statistics for Fare by Pclass:
             mean   median        std
Pclass                               
1       94.280297  60.0000  84.435858
2       22.202104  15.7500  13.991877
3       12.459678   7.8958  10.803698


### Summary

In this notebook, we learn the importance of data transformation and manipulation in data analysis and provide a step-by-step tutorial on how to perform some of the operations using Python's libraries. The key points covered are as follows:

- You can obtain a summary of the dataset using the describe() method. This summary provides statistics such as count, mean, standard deviation, minimum, maximum, and quartiles for numerical columns in the dataset.
- Filtering and subsetting data helps you select specific columns or rows from the dataset based on conditions. The example showcases how to filter passengers older than 30 and those older than 30 in Pclass 1, combining multiple conditions.
- Handling missing data is a common challenge. The example above illustrates how to identify missing values using .isnull().sum() and how to handle them by either removing rows with missing values using .dropna() or filling missing values with a specific value using .fillna().
- Sometimes, new features need to be created to better represent the data or to look at the data from a new angle. The example above shows how to create a new feature, 'Age_Category', by categorizing passengers based on their age into 'Child', 'Adult', or 'Senior'. It also demonstrates the creation of the 'Family_Size'  feature based on existing columns.
- Aggregation and grouping are essential for summarizing data. The example above presents how to group data by a specific column (e.g., 'Pclass') and calculate summary statistics (e.g., mean, median, and standard deviation) for other columns (e.g., 'Age' and 'Fare') within those groups.

That's it! Data transformation and manipulation are fundamental skills for any data analyst or scientist. With Pandas and Python, you have a powerful toolkit at your disposal. Remember to adapt these techniques to your specific data and analysis goals. Thank you for joining me today, and happy data wrangling!